mc/query/compile.go

```diff
@@ -0,0 +1,478 @@
+		sqlq = "SELECT %s FROM Statement JOIN Envelope ON Statement.id = Envelope.id"
+		join = true
+	}
+
+	cols, err := compileQueryColumns(q, join)
+	if err != nil {
+		return "", nil, err
+	}
+	sqlq = fmt.Sprintf(sqlq, cols)
+
+	crit, err := compileQueryCriteria(q, join)
+	if err != nil {
+		return "", nil, err
+	}
+	if crit != "" {
+		sqlq = fmt.Sprintf("%s WHERE %s", sqlq, crit)
+	}
+
+	if q.limit > 0 {
+		sqlq = fmt.Sprintf("%s LIMIT %d", sqlq, q.limit)
+	}
+
+	rsel, err := compileQueryRowSelector(q)
+	if err != nil {
+		return "", nil, err
+	}
+
+	return sqlq, rsel, nil
+}
+
+func compileQueryColumns(q *Query, join bool) (string, error) {
+	switch sel := q.selector.(type) {
+	case SimpleSelector:
+		col := selectorColumn(sel, selectorColumnSimple)
+		return disambigSelector(col, join), nil
+
+	case CompoundSelector:
+		if len(sel) == 1 {
+			col := selectorColumn(sel[0], selectorColumnSimple)
+			return disambigSelector(col, join), nil
+		}
+
+		cols := make([]string, len(sel))
+		for x := 0; x < len(sel); x++ {
+			col := selectorColumn(sel[x], selectorColumnCompound)
+			cols[x] = disambigSelector(col, join)
+		}
+		return strings.Join(cols, ", "), nil
+
+	case *FunctionSelector:
+		col := selectorColumn(sel.sel, selectorColumnFun)
+		return fmt.Sprintf("%s(%s)", sel.op, disambigSelector(col, join)), nil
+
+	default:
+		return "", QueryCompileError(fmt.Sprintf("Unexpected selector type: %T", sel))
+	}
+}
+
+// when we are JOINing, id is ambiguous because it is a column in both tables;
+// this funciton disambiguates
+func disambigSelector(col string, join bool) string {
+	switch {
+	case !join:
+		return col
+	case col == "id":
+		return "Envelope.id"
+	default:
+		return col
+	}
+}
+
+func selectorColumn(sel SimpleSelector, rename map[string]string) string {
+	col, ok := rename[string(sel)]
+	if !ok {
+		col = string(sel)
+	}
+	return col
+}
+
+var selectorColumnSimple = map[string]string{
+	"*":         "data",
+	"body":      "data",
+	"namespace": "DISTINCT namespace",
+	"publisher": "DISTINCT publisher",
+	"source":    "DISTINCT source"}
+
+var selectorColumnCompound = map[string]string{
+	"*":    "data",
+	"body": "data"}
+
+var selectorColumnFun = map[string]string{
+	"*":         "*",
+	"body":      "data",
+	"namespace": "DISTINCT namespace",
+	"publisher": "DISTINCT publisher",
+	"source":    "DISTINCT source"}
+
+func compileQueryCriteria(q *Query, join bool) (string, error) {
+	nscrit := compileNamespaceCriteria(q.namespace)
+	if q.criteria == nil {
+		return nscrit, nil
+	}
+
+	scrit, err := compileSelectorCriteria(q.criteria, join)
+	if err != nil {
+		return "", err
+	}
+
+	if nscrit != "" {
+		return fmt.Sprintf("%s AND %s", nscrit, scrit), nil
+	}
+	return scrit, nil
+}
+
+func compileNamespaceCriteria(ns string) string {
+	switch {
+	case ns == "*":
+		return ""
+	case ns[len(ns)-1] == '*':
+		pre := ns[:len(ns)-2]
+		return fmt.Sprintf("namespace LIKE '%s%%'", pre)
+	default:
+		return fmt.Sprintf("namespace = '%s'", ns)
```

Comment:
Is this safe because we won't be able to get a sql injection payload past the MC query parser?